examples/hermes-engine/Program.cs

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Hermes.Example;
+using Microsoft.JavaScript.NodeApi;
+using static Microsoft.JavaScript.NodeApi.JSNativeApi.Interop;
+
+JSDispatcherQueueController controller = JSDispatcherQueueController.CreateOnDedicatedThread();
+controller.DispatcherQueue.TryEnqueue(() =>
```

Comment:
I'm trying to understand how the dispatcher queue can work, when only Node-API callbacks (not JS callbacks) will use it. For example, consider the following:

```C#
JSValue scriptResult = JSNativeApi.RunScript('new Promise((resolve) => setTimeout(resolve, 1000))');
await ((JSPromise)scriptResult).AsTask();
```

How would the `setTimeout` callback get scheduled (so the Promise gets completed), when the inner callback doesn't get scheduled on the dispatcher queue?

Reply:
The `setTimeout` is not part of the JS engine. It is a part of the hosting environment. For it to work we must add more code to the host. In this case we will need to:
1. Extend the `JSDispatcherQueue` to support timers. It is part of WinRT `DispatcherQueue` - I just did not implement it for now. The implementation typically uses a different `Monitor.Wait` overload with a specified timeout to wake up.
2. Add native implementation for the `setTimeout` polyfill which will create the new timer in the queue in response to this request. 

The way it will work:
- This C# must be exeuted in the `JSDispatcherQueue` task.
- As a part of script execution, the lambda provided to the `Promise` constructor is executed synschonously as a part of the `Promise` constructor.
- The lambda calls the `setTimeout` polyfill function which has C# impleemnt that will schedule a timer job in the queue.
- When the timeout is done it will complete the JS `Promise` which in turn will fulfill the C# `JSPromise` which will continue the execution.

Note that the `JSDispatcherQueue` and the `SynchronizationContext` are the same queue in this scenario.

I should implement support for the async scenarios as the next step to produce more intersting example.

Reply:
OK. That clarifies it for me, thanks. `setTimeout` was just an example, but I guess anything that could possibly schedule future work must be implemented by the host.

Reply:
Yes, you are right. I want to add support for timer in the next PR. This PR is already too big and, in some sense, has one piece of complete functionality.